packages/morphic-adt/src/matcher.ts

```diff
@@ -98,6 +120,16 @@ export interface Matchers<A, Tag extends keyof A> {
 export const Matchers = <A, Tag extends keyof A>(tag: Tag) => (keys: KeysDefinition<A, Tag>): Matchers<A, Tag> => {
   const inKeys = isIn(keys)
   const match = (match: any, def?: any) => (a: any): any => (match[a[tag]] || def)(a)
+  const matchLens = (cases: any) =>
+    new Lens(
+      (s: any) => cases[s[tag]].get(s),
+      (a: any) => (s: any) => cases[s[tag]].set(a)(s)
+    )
+  const matchOptional = (cases: any) =>
+    new Optional(
+      (s: any) => (s[tag] in cases ? cases[s[tag]].getOption(s) : none),
+      (a: any) => (s: any) => (s[tag] in cases ? cases[s[tag]].set(a)(s) : s)
```

Comment:
please, factorise the s[tag] access 
also reduce the cost of `s[tag] in cases` and `cases[s[tag]]` by just having a single access